File: content/snippets/typescript/language_basics/stacks_and_queues.ts
#!/usr/bin/env ts-node

class Stack<T> {
  private items: T[] = [];
  push(v: T): void { this.items.push(v); }
  pop(): void { this.items.pop(); }
  top(): T { return this.items[this.items.length - 1]; }
  isEmpty(): boolean { return this.items.length === 0; }
}

class Queue<T> {
  private items: T[] = [];
  enqueue(v: T): void { this.items.push(v); }
  dequeue(): void { this.items.shift(); }
  front(): T { return this.items[0]; }
  isEmpty(): boolean { return this.items.length === 0; }
}

function main(): void {
  const stk = new Stack<number>();
  for (const v of [1, 2, 3]) stk.push(v);
  console.log("stack top:", stk.top());  // 3 (LIFO)
  stk.pop();
  console.log("after pop:", stk.top());  // 2

  const q = new Queue<number>();
  for (const v of [1, 2, 3]) q.enqueue(v);
  console.log("queue front:", q.front());  // 1 (FIFO)
  q.dequeue();
  console.log("after dequeue:", q.front());  // 2
}

main();

File: content/snippets/cpp/language_basics/stacks_and_queues.cpp
#include <iostream>
#include <queue>
#include <stack>

int main() {
    std::stack<int> stk;
    for (int v : {1, 2, 3}) stk.push(v);
    std::cout << "stack top: " << stk.top() << "\n";  // 3 (LIFO)
    stk.pop();
    std::cout << "after pop: " << stk.top() << "\n";  // 2

    std::queue<int> q;
    for (int v : {1, 2, 3}) q.push(v);
    std::cout << "queue front: " << q.front() << "\n";  // 1 (FIFO)
    q.pop();
    std::cout << "after dequeue: " << q.front() << "\n";  // 2
    return 0;
}
